packages/custom_lint/lib/src/workspace.dart

```diff
@@ -633,39 +645,80 @@ publish_to: 'none'
               .nonNulls
               .toList(),
         ),
-    };
+    }.map(
+      (key, value) {
+        // Any dependency which is in both `dependencies` and `dev_dependencies` is treated as a regular dependency.
+        if (value.regularDependencies.isNotEmpty &&
+            value.devDependencies.isNotEmpty) {
+          value.regularDependencies.addAll(value.devDependencies);
+          value.devDependencies.clear();
+        }
+        return MapEntry(key, value);
+      },
+    );
 
-    // A flag for whether the "dependencies:" header has been written.
-    var didWriteDevDependenciesHeader = false;
-    // Write dev_dependencies
+    final dependencyConstraints = (
+      dependencies: <String, String>{},
+      devDependencies: <String, String>{},
+    );
+
+    // Iterate over each plugin and compute their constraints.
     for (final name in uniquePluginNames) {
       final allDependencies = dependenciesByName[name];
       if (allDependencies == null) continue;
 
-      // // We don't write dev_dependencies which are sometimes prod dependencies,
-      // // as dev_dependencies, because then we'd be specifying the dependency twice.
-      if (
-          // allDependencies.dependencies.isNotEmpty ||
-          allDependencies.devDependencies.isEmpty) {
+      // Determine if this is a dev_dependency or a dependency.
+      // Note: Dependencies can't be in both `dependencies` and `dev_dependencies`.
+      // So these isNonEmpty checks mutually exclusive.
+      final bool isDevDependency;
```

Comment:
We should preserve the source of a dependency.
If it's from dependencies, we write it as dependencies. If it is a dev_dependency, we write it as so.
And the constraints should match.

Reply:
But what if it's in both?

e.g.
```bash
.
├── b
│   └── pubspec.yaml # drift_dev in dev_dependencies
├── c
│   └── pubspec.yaml # drift_dev in dependencies
└── a
    └── pubspec.yaml # Root Package
```
Project `A` depends on package `B` and `C`.
If `drift_dev`is a `dev_dependencies` in package `B` and a regular `dependency` in `C`, how should it be written? 

I spent a while trying to understand what this code is doing. If my assumptions are wrong please enlighten me

Reply:
Actually, I think we should always be writing plugins to `dependencies:` , not `dev_dependencies:`

Because plugins use them directly rather than through command. If the linter was running against its own source, we'd probably see a warning that plugins should be in "dependencies" instead of dev_dependencies.

It works regardless because Dart sometimes doesn't care. But the current code is probably not right.

Reply:
>Because plugins use them directly rather than through command. If the linter was running against its own source, we'd probably see a warning that plugins should be in "dependencies" instead of dev_dependencies. 

Good point.

I've updated the code to combine `dev_dependencies` and `dependencies` and to always write them as regular `dependencies` in the generated pubspec. I've updated the tests for this too.

Reply:
I've also added a test for this specific scenario.

Reply:
Sorry, let me have another look.